src/databricks/labs/ucx/source_code/astlinter.py

```diff
@@ -0,0 +1,58 @@
+import ast
+import logging
+
+logger = logging.getLogger(__name__)
+
+
+class MatchingVisitor(ast.NodeVisitor):
+
+    def __init__(self, node_type: type, match_nodes: list[tuple[str, type]]):
+        self.matched_nodes: list[ast.AST] = []
+        self._node_type = node_type
+        self._match_nodes = match_nodes
+
+    # pylint: disable=invalid-name
+    def visit_Call(self, node: ast.Call):
+        if self._node_type is not ast.Call:
+            return
+        try:
+            if self._matches(node.func, 0):
+                self.matched_nodes.append(node)
+        except NotImplementedError as e:
+            logger.warning(f"Missing implementation: {e.args[0]}")
+
+    def _matches(self, node: ast.AST, depth: int):
+        if depth >= len(self._match_nodes):
+            return False
+        pair = self._match_nodes[depth]
+        if not isinstance(node, pair[1]):
+            return False
+        next_node: ast.AST | None = None
+        if isinstance(node, ast.Attribute):
+            if node.attr != pair[0]:
+                return False
+            next_node = node.value
+        elif isinstance(node, ast.Name):
+            if node.id != pair[0]:
+                return False
+        else:
+            raise NotImplementedError(str(type(node)))
+        if next_node is None:
+            # is this the last node to match ?
+            return len(self._match_nodes) - 1 == depth
+        return self._matches(next_node, depth + 1)
+
+
+class ASTLinter:
+
+    def __init__(self):
+        self._module: ast.Module | None = None
+
+    def parse(self, code: str):
+        self._module = ast.parse(code)
```

Comment:
```suggestion
    def __init__(self, code: str):
        self._module = ast.parse(code)
```
so that we either fail initialising or have a valid state.

Reply:
I respectfully disagree. A constructor should not perform processing, especially not when that processing may fail under uncontrolled conditions (source code received from outside)

Reply:
Changed to a factory pattern

Reply:
actually the reason for this is that ASTLinter is a base class for PythonLinter, which also needs to follow Linter conventions... changing that now

Reply:
done